middleman_protocol/tests/tests.py

```diff
@@ -0,0 +1,113 @@
+    def test_that_abstract_middleman_message_factory_with_different_payload_types_should_create_proper_middleman_message(
+        self,
+        expected_middleman_message_type,
+        payload_type,
+        payload,
+    ):
+        message = AbstractMiddlemanMessage.factory(
+            payload_type,
+            payload,
+            self.request_id,
+        )
+
+        assertpy.assert_that(message).is_instance_of(expected_middleman_message_type)
+        assertpy.assert_that(message.payload_type).is_equal_to(payload_type)
+
+    def test_that_abstract_middleman_message_instantiation_should_raise_exception(self):
+        with pytest.raises(NotImplementedError):
+            AbstractMiddlemanMessage(
+                Ping(),
+                self.request_id,
+            )
+
+
+class TestSerializeMessagesFromMiddlemanProtocol:
+
+    request_id = 99
+
+    @pytest.mark.parametrize(('middleman_message_type', 'payload'), [
+        (GolemMessageMiddlemanMessage,            Ping()),
+        (ErrorMiddlemanMessage,                   ('error_message', 'error.code')),
+        (AuthenticationChallengeMiddlemanMessage, b'random_bytes'),
+        (AuthenticationResponseMiddlemanMessage,  b'TODO'),
+    ])  # pylint: disable=no-self-use
+    def test_that_serializing_and_deserializing_message_should_preserve_original_data(
+        self,
+        middleman_message_type,
+        payload,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+        raw_message = middleman_message.serialize()
+        deserialized_message = AbstractMiddlemanMessage.deserialize(raw_message)
+
+        assert isinstance(deserialized_message, type(payload))
+        assert deserialized_message == payload
+
+
+class TestTransferMessagesFromMiddlemanProtocol:
+
+    request_id = 99
+
+    @pytest.mark.parametrize(('middleman_message_type', 'payload'), [
+        (GolemMessageMiddlemanMessage,            Ping()),
+        (ErrorMiddlemanMessage,                   ('error_message', 'error.code')),
+        (AuthenticationChallengeMiddlemanMessage, b'random_bytes'),
+        (AuthenticationResponseMiddlemanMessage,  b'TODO'),
+    ])  # pylint: disable=no-self-use
+    def test_that_sending_message_over_tcp_socket_should_preserve_original_data(
+        self,
+        middleman_message_type,
+        payload,
+    ):
+        middleman_message = middleman_message_type(payload, self.request_id)
+        raw_message = middleman_message.serialize()
+
+        with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as server_socket:
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as client_socket:
+                server_socket.bind(('127.0.0.1', 8001))
+                server_socket.listen(1)
+
+                client_socket.connect(('127.0.0.1', 8001))
+
+                connection, address = server_socket.accept()
+
+                client_socket.send(raw_message)
+
+                received_data = b''
```

Comment:
Please make a helper generator function that will accept some form of socket reader (it would be a wrapper for `socket` or `asyncio`) so it can be (re)used in SS and MM ;)

Reply:
I am not sure if this is possible when we want it to handle regular and async socket receiving at the same time. As I understand, this helper should be wrapped in `async` keyword and `await` from reader for async socket, but it won't work for regular socket.

Reply:
You're right. The fact that asyncio can't easily be mixed with synchronous code is a known problem. See [Controlling Python Async Creep](https://hackernoon.com/controlling-python-async-creep-ec0a0f4b79ba) for some ideas but I see that this guy is suggesting using threads which I don't think is a good idea here. But I only skimmed it so maybe there's something good in there.

Anyway, if you can't come up with anything better, just create separate sync and async helpers. And the async one can be created later, in one of the tasks related to Middleman.

Reply:
I have read it and it is done there as I have predicted, it is a function that have two separate flows for regular and async calls :) So I will add two separate helpers.

Reply:
:+1: done